private/bufpkg/bufwasm/bufwasm.go

```diff
@@ -0,0 +1,253 @@
+	}()
+	// Note: before we start accepting user plugins, we should do more
+	// validation on the metadata here: file path cleaning etc.
+	compiledModule, err := runtime.CompileModule(ctx, plugin)
+	if err != nil {
+		return nil, fmt.Errorf("error compiling wasm: %w", err)
+	}
+	var bufsectionBytes []byte
+	for _, section := range compiledModule.CustomSections() {
+		if section.Name() == CustomSectionName {
+			bufsectionBytes = append(bufsectionBytes, section.Data()...)
+		}
+	}
+	compiledPlugin := &CompiledPlugin{Module: compiledModule}
+	if len(bufsectionBytes) > 0 {
+		metadata := &wasmpluginv1.Meta{}
+		if err := proto.Unmarshal(bufsectionBytes, metadata); err != nil {
+			return nil, err
+		}
+		compiledPlugin.Metadata = metadata
+	}
+	return compiledPlugin, nil
+}
+
+// Run executes a plugin. If the plugin exited with non-zero status, this
+// returns a *PluginExecutionError.
+func (e *WASMPluginExecutor) Run(
+	ctx context.Context,
+	plugin *CompiledPlugin,
+	stdin io.Reader,
+	stdout io.Writer,
+) (retErr error) {
+	name := plugin.Module.Name()
+	if name == "" {
+		// Some plugins will attempt to read argv[0], but don't have a
+		// name in the wasm file. Fallback to this.
+		name = "protoc-gen-wasm"
+	}
+
+	stderr := bytes.NewBuffer(nil)
+	config := wazero.NewModuleConfig().
+		WithName(plugin.Module.Name()).
+		WithArgs(append([]string{name}, plugin.Metadata.GetArgs()...)...).
+		WithStdin(stdin).
+		WithStdout(stdout).
+		WithStderr(stderr)
+	if len(plugin.Metadata.GetFiles()) > 0 {
+		mapFS := make(fstest.MapFS, len(plugin.Metadata.GetFiles()))
+		for _, file := range plugin.Metadata.Files {
+			mapFS[strings.TrimPrefix(file.Path, "/")] = &fstest.MapFile{
+				Data: file.Contents,
+			}
+		}
+		config = config.WithFS(mapFS)
+	}
+
+	runtime := wazero.NewRuntimeWithConfig(ctx, e.runtimeConfig)
+	defer func() {
+		retErr = multierr.Append(retErr, runtime.Close(ctx))
+	}()
+
+	var err error
+	switch plugin.ABI() {
+	case wasmpluginv1.WasmABI_WASM_ABI_GOJS:
+		hostModuleBuilder := runtime.NewHostModuleBuilder("go")
+		gojs.NewFunctionExporter().ExportFunctions(hostModuleBuilder)
+		var module api.Module
+		module, err = hostModuleBuilder.Instantiate(ctx)
+		if err != nil {
+			return fmt.Errorf("error instantiating gojs: %w", err)
+		}
+		defer func() {
+			retErr = multierr.Append(retErr, module.Close(ctx))
+		}()
+		err = gojs.Run(ctx, runtime, plugin.Module, gojs.NewConfig(config))
+	case wasmpluginv1.WasmABI_WASM_ABI_WASI_SNAPSHOT_PREVIEW1:
+		var closer api.Closer
+		closer, err = wasi_snapshot_preview1.NewBuilder(runtime).Instantiate(ctx)
+		if err != nil {
+			return fmt.Errorf("error instantiating wasi: %w", err)
+		}
+		defer func() {
+			retErr = multierr.Append(retErr, closer.Close(ctx))
+		}()
+		var module api.Module
+		module, err = runtime.InstantiateModule(ctx, plugin.Module, config)
+		defer func() {
+			retErr = multierr.Append(retErr, module.Close(ctx))
+		}()
+	default:
```

Comment:
Are you missing a `wasi.Run` or something? I'm not sure where the plugin actually gets executed in this block.